File: scala/src/main/scala/a_lists/P02.scala
package a_lists

import java.util.NoSuchElementException
import java.lang.IndexOutOfBoundsException
// P02. Find the last but one element of a list.
object P02 {

  def penultimate[A](xs: List[A]): A = {
    if(xs.length < 2) {
      throw new IndexOutOfBoundsException
    }
    xs(xs.length-2)
  }

  def penultimateRecursive[A](xs: List[A]): A = xs match {
    case x :: y :: Nil => x
    case _ :: tail => penultimateRecursive(tail)
    case _         => throw new NoSuchElementException
  }

}

File: ruby/a_lists/p02.rb
class P02

  # Find the next-to-last element in a list

  def initialize(list)
    @list = list || []
  end

  def next_to_last_element
    @list[-2]
  end

  def last_by_length_index
    @list[@list.length - 2] 
  end

end